### PR TITLE
Changes in the SLQ Algorithm

### PR DESCRIPTION
**Pull request by karthikrangasai**
I've made the changes for the Rademacher Random Vector
Changed the Lanczos Algorithm as required and also updated the slq function

Added unit test in test/slq.jl

### CODE REVIEW

File: src/slq.jl [hunk 1]
```diff
@@ -12,65 +12,65 @@ end
...
 
-        Y = eigvecs(T)
-        theta = eigvals(T)
+            Y = eigvecs(T)
+            theta = eigvals(T)
 
-        for k = 1:size(theta,1)
-            trace = trace + ((basisVec(size(Y[:,k],1),1))' * Y[:,k])^2 * f(theta[k])
+            for k = 1:size(theta,1)
+                trace = trace + ((basisVec(size(Y[:,k],1),1))' * Y[:,k])^2 * f(theta[k])
```
Comment (mohdibntarek) [inline src/slq.jl]: `trace = trace + Y[1,k]^2 * f(theta[k])`

File: src/slq.jl [hunk 2]
```diff
@@ -12,65 +12,65 @@ end
...
+        mul!(r, A, q)
+        r .-= Beta[j-1] .* v
         alpha = q' * r
+
         r = r - alpha .* q
-        beta = norm(r)
-        if beta == 0
-            break
-        end
+        # beta = norm(r)
```
Comment (mohdibntarek) [inline src/slq.jl]: Use inplace version `r .= r .- alpha .* q`

File: src/slq.jl [hunk 3]
```diff
@@ -12,65 +12,65 @@ end
...
 function randomRademacherVector(x::Int64)
-    mp = [-1 1];
-    v = rand(1,x);
-
-    for i = range(1, length = x)
-        v[i] = mp[(v[i] < 0.5) + 1];
-    end
-
-    return v
+    return rand(-1.0:2.0:1.0, x)
```
Comment (mohdibntarek) [inline src/slq.jl]: Allow passing a number type `T` and return a random vector with that type as the `eltype`. For example:
```julia
o = one(T)
t = 2*o
rand(-o:t:o, x)
```
This enables this algorithm to generalize to `Float32`, higher precision types, etc.

File: test/slq.jl [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+eps = rand(1)
+neta = rand(1)
+f(x) = exp(x)
+g(x) = x^2
+
+@testset "Testing Stochastic Lanczos Quadrature" begin
+    @testset "Testing for Dense Matrices" begin
+        @testset "For Symmetric Matrices of size ($m, $n) with analytic function f(x) = exp(x)" for
+            (m, n) in ((50, 50), (100, 100), (500, 500), (1000, 1000), (5000, 5000))
```
Comment (mohdibntarek) [inline test/slq.jl]: No need for `m` and `n`, just one is enough.

File: test/slq.jl [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            A = rand(m, n)
+            testMatrix = 0.5 * (A + A')
+            eigMax = eigmax(testMatrix)
+            eigMin = eigmin(testMatrix)
+            condNum = eigMax/eigMin
+            Mp = f(eigMax)
+            mp = f(eigMin)
+            K = ((eigMax - eigMin) * ((sqrt(condNum) - 1)^2) * Mp)/(sqrt(condNum) * mp)
+
+            m = ceil((sqrt(condNum) * log(K/eps)) / 4)
```
Comment (mohdibntarek) [inline test/slq.jl]: Perhaps compute `m` and `nv` defaults inside the `slq` function?

File: test/slq.jl [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            nv = ceil((24 * log(2/neta)) / (eps^2))
+
+            @time computed = slq(testMatrix, f, m, nv)
+            @time actual = tr(inv(testMatrix))
+            @test error(computed, actual)
```
Comment (mohdibntarek) [inline test/slq.jl]: Use `isapprox`

**Review by mohdibntarek — CHANGES_REQUESTED** [review]
Looks good! I left some comments. Would also be nice to test the algorithm with matrices that have a large condition number. Also make sure to add the test dependencies including stdlib packages in the Project.toml file. To add packages to Project.toml, follow the steps in https://discourse.julialang.org/t/convert-require-to-project-toml-and-manifest-toml/17775/2.

File: test/slq.jl [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    else
+        return false
+    end
+end
+
+function signCheck(eMax::Float64, eMin::Float64)
+    if sign(eMax) == sign(eMin)
+        return true
+    else
+        retuen false
```
Comment (mohdibntarek) [inline test/slq.jl]: `return`

**Comment by mohdibntarek** [comment]
I just ran the tests on my machine. They error. Please make sure the tests pass in REPL, and then setup Travis for automatic testing of the PR.

File: test/slq.jl (hunk 5 above)
Comment (karthikrangasai) [reply test/slq.jl]: `m` and `nv` are inputs that user gives, so I'm computing them before hand all the time

Comment (mohdibntarek) [reply test/slq.jl]: Yes, but it would be nice to have reasonable default values.

**Pull request merged**